From d2da03b7939def1ed84aa7061b40964ac10139ef Mon Sep 17 00:00:00 2001
From: Asriel Dreemurr <asriel.danctnix@gmail.com>
Date: Sun, 18 Oct 2020 00:31:45 +0700
Subject: [PATCH 01/19] (halium) init: start inside LXC container without
 SELinux

amartinz: keep first stage init setup for recovery, otherwise
recovery will not boot.

Change-Id: Ief7e2d5c20bec6c83e946f26c0285e1b0822b0c1
Signed-off-by: Alexander Martinz <alex@amartinz.at>
Signed-off-by: Eugenio Paolantonio (g7) <me@medesimo.eu>
---
 init/Android.bp           |  2 ++
 init/first_stage_init.cpp | 44 ++++++++++++++++++++++++++++-----------
 init/init.cpp             |  7 ++++++-
 init/property_service.cpp | 23 +++++++++++++-------
 init/service.cpp          | 15 +++++++------
 init/subcontext.cpp       |  3 ++-
 init/util.cpp             | 18 +++++++++-------
 7 files changed, 78 insertions(+), 34 deletions(-)

diff --git a/init/Android.bp b/init/Android.bp
index 327c56e..2d747ea 100644
--- a/init/Android.bp
+++ b/init/Android.bp
@@ -403,6 +403,7 @@ init_first_stage_cc_defaults {
         "-Wall",
         "-Wextra",
         "-Wno-unused-parameter",
+	"-Wno-unused-function",
         "-Werror",
         "-DALLOW_FIRST_STAGE_CONSOLE=0",
         "-DALLOW_LOCAL_PROP_OVERRIDE=0",
@@ -585,6 +586,7 @@ cc_defaults {
         "-Wall",
         "-Wextra",
         "-Wno-unused-parameter",
+	"-Wno-unused-function",
         "-Werror",
     ],
     static_libs: [
diff --git a/init/first_stage_init.cpp b/init/first_stage_init.cpp
index c4d0f75..1a521c6 100644
--- a/init/first_stage_init.cpp
+++ b/init/first_stage_init.cpp
@@ -330,14 +330,22 @@ int FirstStageMain(int argc, char** argv) {
     CHECKCALL(setenv("PATH", _PATH_DEFPATH, 1));
     // Get the basic filesystem setup we need put together in the initramdisk
     // on / and then we'll let the rc file figure out the rest.
-    CHECKCALL(mount("tmpfs", "/dev", "tmpfs", MS_NOSUID, "mode=0755"));
-    CHECKCALL(mkdir("/dev/pts", 0755));
-    CHECKCALL(mkdir("/dev/socket", 0755));
-    CHECKCALL(mkdir("/dev/dm-user", 0755));
-    CHECKCALL(mount("devpts", "/dev/pts", "devpts", 0, NULL));
+    // We need to disable certain things for Halium while the recovery needs it.
+    bool is_recovery = IsRecoveryMode();
+
+    // Disabled for Halium, mounted by host system instead
+    if (!is_recovery) {
+        mkdir("/dev/socket", 0755);
+    } else {
+        CHECKCALL(mount("tmpfs", "/dev", "tmpfs", MS_NOSUID, "mode=0755"));
+        CHECKCALL(mkdir("/dev/pts", 0755));
+        CHECKCALL(mkdir("/dev/socket", 0755));
+        CHECKCALL(mount("devpts", "/dev/pts", "devpts", 0, NULL));
 #define MAKE_STR(x) __STRING(x)
-    CHECKCALL(mount("proc", "/proc", "proc", 0, "hidepid=2,gid=" MAKE_STR(AID_READPROC)));
+	CHECKCALL(mount("proc", "/proc", "proc", 0, "hidepid=2,gid=" MAKE_STR(AID_READPROC)));
 #undef MAKE_STR
+    }
+    CHECKCALL(mkdir("/dev/dm-user", 0755));
     // Don't expose the raw commandline to unprivileged processes.
     CHECKCALL(chmod("/proc/cmdline", 0440));
     std::string cmdline;
@@ -348,8 +356,11 @@ int FirstStageMain(int argc, char** argv) {
     android::base::ReadFileToString("/proc/bootconfig", &bootconfig);
     gid_t groups[] = {AID_READPROC};
     CHECKCALL(setgroups(arraysize(groups), groups));
-    CHECKCALL(mount("sysfs", "/sys", "sysfs", 0, NULL));
-    CHECKCALL(mount("selinuxfs", "/sys/fs/selinux", "selinuxfs", 0, NULL));
+    // Disabled for Halium
+    if (is_recovery) {
+        CHECKCALL(mount("sysfs", "/sys", "sysfs", 0, NULL));
+        CHECKCALL(mount("selinuxfs", "/sys/fs/selinux", "selinuxfs", 0, NULL));
+    }
 
     CHECKCALL(mknod("/dev/kmsg", S_IFCHR | 0600, makedev(1, 11)));
 
@@ -361,8 +372,14 @@ int FirstStageMain(int argc, char** argv) {
     CHECKCALL(mknod("/dev/urandom", S_IFCHR | 0666, makedev(1, 9)));
 
     // This is needed for log wrapper, which gets called before ueventd runs.
-    CHECKCALL(mknod("/dev/ptmx", S_IFCHR | 0666, makedev(5, 2)));
-    CHECKCALL(mknod("/dev/null", S_IFCHR | 0666, makedev(1, 3)));
+    // Can be created by LXC on Halium
+    if (is_recovery) {
+        CHECKCALL(mknod("/dev/ptmx", S_IFCHR | 0666, makedev(5, 2)));
+        CHECKCALL(mknod("/dev/null", S_IFCHR | 0666, makedev(1, 3)));
+    } else {
+        mknod("/dev/ptmx", S_IFCHR | 0666, makedev(5, 2));
+        mknod("/dev/null", S_IFCHR | 0666, makedev(1, 3));
+    }
 
     // These below mounts are done in first stage init so that first stage mount can mount
     // subdirectories of /mnt/{vendor,product}/.  Other mounts, not required by first stage mount,
@@ -397,7 +414,8 @@ int FirstStageMain(int argc, char** argv) {
         for (const auto& [error_string, error_errno] : errors) {
             LOG(ERROR) << error_string << " " << strerror(error_errno);
         }
-        LOG(FATAL) << "Init encountered errors starting first stage, aborting";
+	// Halium: log errors but do not abort
+        // LOG(FATAL) << "Init encountered errors starting first stage, aborting";
     }
 
     LOG(INFO) << "init first stage started!";
@@ -544,7 +562,9 @@ int FirstStageMain(int argc, char** argv) {
            1);
 
     const char* path = "/system/bin/init";
-    const char* args[] = {path, "selinux_setup", nullptr};
+    // Halium: skip selinux_setup
+    // const char* args[] = {path, "selinux_setup", nullptr};
+    const char* args[] = {path, "second_stage", nullptr};
     auto fd = open("/dev/kmsg", O_WRONLY | O_CLOEXEC);
     dup2(fd, STDOUT_FILENO);
     dup2(fd, STDERR_FILENO);
diff --git a/init/init.cpp b/init/init.cpp
index aeccd66..3fb5f7e 100644
--- a/init/init.cpp
+++ b/init/init.cpp
@@ -1078,7 +1078,9 @@ int SecondStageMain(int argc, char** argv) {
     // Queue an action that waits for coldboot done so we know ueventd has set up all of /dev...
     am.QueueBuiltinAction(wait_for_coldboot_done_action, "wait_for_coldboot_done");
     // ... so that we can start queuing up actions that require stuff from /dev.
-    am.QueueBuiltinAction(SetMmapRndBitsAction, "SetMmapRndBits");
+    // Disabled for Halium
+    //am.QueueBuiltinAction(MixHwrngIntoLinuxRngAction, "MixHwrngIntoLinuxRng");
+    //am.QueueBuiltinAction(SetMmapRndBitsAction, "SetMmapRndBits");
     Keychords keychords;
     am.QueueBuiltinAction(
             [&epoll, &keychords](const BuiltinArguments& args) -> Result<void> {
@@ -1093,6 +1095,9 @@ int SecondStageMain(int argc, char** argv) {
     // Trigger all the boot actions to get us started.
     am.QueueEventTrigger("init");
 
+    // Disabled for Halium
+    // am.QueueBuiltinAction(MixHwrngIntoLinuxRngAction, "MixHwrngIntoLinuxRng");
+
     // Don't mount filesystems or start core system services in charger mode.
     std::string bootmode = GetProperty("ro.bootmode", "");
     if (bootmode == "charger") {
diff --git a/init/property_service.cpp b/init/property_service.cpp
index b3985f8..fba383c 100644
--- a/init/property_service.cpp
+++ b/init/property_service.cpp
@@ -160,7 +160,8 @@ void StopSendingMessages() {
 }
 
 bool CanReadProperty(const std::string& source_context, const std::string& name) {
-    const char* target_context = nullptr;
+    // Disabled for Halium
+    /*const char* target_context = nullptr;
     property_info_area->GetPropertyInfo(name.c_str(), &target_context, nullptr);
 
     PropertyAuditData audit_data;
@@ -172,12 +173,14 @@ bool CanReadProperty(const std::string& source_context, const std::string& name)
 
     auto lock = std::lock_guard{selinux_check_access_lock};
     return selinux_check_access(source_context.c_str(), target_context, "file", "read",
-                                &audit_data) == 0;
+                                &audit_data) == 0;*/
+    return true;
 }
 
 static bool CheckMacPerms(const std::string& name, const char* target_context,
                           const char* source_context, const ucred& cr) {
-    if (!target_context || !source_context) {
+    // Disabled for Halium
+    /*if (!target_context || !source_context) {
         return false;
     }
 
@@ -189,6 +192,9 @@ static bool CheckMacPerms(const std::string& name, const char* target_context,
     auto lock = std::lock_guard{selinux_check_access_lock};
     return selinux_check_access(source_context, target_context, "property_service", "set",
                                 &audit_data) == 0;
+    */
+
+    return true;
 }
 
 void NotifyPropertyChange(const std::string& name, const std::string& value) {
@@ -635,11 +641,13 @@ static void handle_property_set_fd(int fd) {
         prop_name[PROP_NAME_MAX-1] = 0;
         prop_value[PROP_VALUE_MAX-1] = 0;
 
-        std::string source_context;
+        std::string source_context = "";
+        // Disabled for Halium
+        /*if (!socket.GetSourceContext(&source_context)) {
         if (!socket.GetSourceContext(&source_context)) {
             PLOG(ERROR) << "Unable to set property '" << prop_name << "': getpeercon() failed";
             return;
-        }
+        }*/
 
         const auto& cr = socket.cred();
         std::string error;
@@ -663,11 +671,12 @@ static void handle_property_set_fd(int fd) {
         }
 
         std::string source_context;
-        if (!socket.GetSourceContext(&source_context)) {
+        // Disabled for Halium
+        /*if (!socket.GetSourceContext(&source_context)) {
             PLOG(ERROR) << "Unable to set property '" << name << "': getpeercon() failed";
             socket.SendUint32(PROP_ERROR_PERMISSION_DENIED);
             return;
-        }
+        }*/
 
         // HandlePropertySet takes ownership of the socket if the set is handled asynchronously.
         const auto& cr = socket.cred();
diff --git a/init/service.cpp b/init/service.cpp
index e7d710b..38521a9 100644
--- a/init/service.cpp
+++ b/init/service.cpp
@@ -75,7 +75,8 @@ using android::base::WriteStringToFile;
 namespace android {
 namespace init {
 
-static Result<std::string> ComputeContextFromExecutable(const std::string& service_path) {
+// Halium: Disabled
+/*static Result<std::string> ComputeContextFromExecutable(const std::string& service_path) {
     std::string computed_context;
 
     char* raw_con = nullptr;
@@ -114,7 +115,7 @@ static Result<std::string> ComputeContextFromExecutable(const std::string& servi
         return Error() << "Could not get process context";
     }
     return computed_context;
-}
+}*/
 
 static bool ExpandArgsAndExecv(const std::vector<std::string>& args, bool sigstop) {
     std::vector<std::string> expanded_args;
@@ -243,9 +244,10 @@ void Service::SetProcessAttributesAndCaps(InterprocessFifo setsid_finished) {
     }
 
     if (!seclabel_.empty()) {
-        if (setexeccon(seclabel_.c_str()) < 0) {
+        // Disabled for Halium
+        /*if (setexeccon(seclabel_.c_str()) < 0) {
             PLOG(FATAL) << "cannot setexeccon('" << seclabel_ << "') for " << name_;
-        }
+        }*/
     }
 
     if (capabilities_) {
@@ -626,7 +628,8 @@ Result<void> Service::Start() {
     }
 
     std::string scon;
-    if (!seclabel_.empty()) {
+    // Disabled for Halium
+    /*if (!seclabel_.empty()) {
         scon = seclabel_;
     } else {
         auto result = ComputeContextFromExecutable(args_[0]);
@@ -634,7 +637,7 @@ Result<void> Service::Start() {
             return result.error();
         }
         scon = *result;
-    }
+    }*/
 
     if (!mount_namespace_.has_value()) {
         // remember from which mount namespace the service should start
diff --git a/init/subcontext.cpp b/init/subcontext.cpp
index 6a095fb..fe3ee4e 100644
--- a/init/subcontext.cpp
+++ b/init/subcontext.cpp
@@ -214,7 +214,8 @@ void Subcontext::Fork() {
 
         // We don't switch contexts if we're running the unit tests.  We don't use std::optional,
         // since we still need a real context string to pass to the builtin functions.
-        if (context_ != kTestContext) {
+        // Disabled in LXC container
+        if (false /*context_ != kTestContext*/) {
             if (setexeccon(context_.c_str()) < 0) {
                 PLOG(FATAL) << "Could not set execcon for '" << context_ << "'";
             }
diff --git a/init/util.cpp b/init/util.cpp
index e760a59..2f047d7 100644
--- a/init/util.cpp
+++ b/init/util.cpp
@@ -93,18 +93,20 @@ Result<uid_t> DecodeUid(const std::string& name) {
  */
 Result<int> CreateSocket(const std::string& name, int type, bool passcred, bool should_listen,
                          mode_t perm, uid_t uid, gid_t gid, const std::string& socketcon) {
-    if (!socketcon.empty()) {
+    // Disabled for Halium
+    /*if (!socketcon.empty()) {
         if (setsockcreatecon(socketcon.c_str()) == -1) {
             return ErrnoError() << "setsockcreatecon(\"" << socketcon << "\") failed";
         }
-    }
+    }*/
 
     android::base::unique_fd fd(socket(PF_UNIX, type, 0));
     if (fd < 0) {
         return ErrnoError() << "Failed to open socket '" << name << "'";
     }
 
-    if (!socketcon.empty()) setsockcreatecon(nullptr);
+    // Disabled for Halium
+    /*if (!socketcon.empty()) setsockcreatecon(nullptr);*/
 
     struct sockaddr_un addr;
     memset(&addr, 0 , sizeof(addr));
@@ -115,10 +117,11 @@ Result<int> CreateSocket(const std::string& name, int type, bool passcred, bool
         return ErrnoError() << "Failed to unlink old socket '" << name << "'";
     }
 
-    std::string secontext;
+    // Disabled for Halium
+    /*std::string secontext
     if (SelabelLookupFileContext(addr.sun_path, S_IFSOCK, &secontext) && !secontext.empty()) {
         setfscreatecon(secontext.c_str());
-    }
+    }*/
 
     if (passcred) {
         int on = 1;
@@ -130,9 +133,10 @@ Result<int> CreateSocket(const std::string& name, int type, bool passcred, bool
     int ret = bind(fd.get(), (struct sockaddr*)&addr, sizeof(addr));
     int savederrno = errno;
 
-    if (!secontext.empty()) {
+    // Disabled for Halium
+    /*if (!secontext.empty()) {
         setfscreatecon(nullptr);
-    }
+    }*/
 
     auto guard = android::base::make_scope_guard([&addr] { unlink(addr.sun_path); });
 
-- 
2.45.2

